### PR TITLE
[IMPROVE] Apply skeleton for message images and on workspace avatar (#4810)

### PR DESCRIPTION
**Pull request by floels**
## Proposed changes

The goal of this PR is to implement the improvement mentioned in issue #4810, namely to display a skeleton placeholder when various images in the app are loading.

The approach I followed is roughly the one described [in the issue](https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/4810#issuecomment-1463993988).

This PR is reviewed commit by commit.

## Issue(s)	

#4810

## How to test or reproduce

1. Clone [this branch made for QA](https://github.com/floels/Rocket.Chat.ReactNative/tree/FOR-QA/IMPROVE/4810-apply-skeleton-images) (sub-branch of this merge request's branch). I tweaked the `handleLoadEnd` method of the new `<ImageSkeletonPlaceholder/>` component to make placeholder skeletons visible for three seconds after the image has loaded. I also had to point to an arbitrary image to check the behavior for the server avatar of `open.rocket.chat` because it didn't show in my local environment, I'm not sure if this is a general issue or not.

2. Move to the cloned branch. Build and launch the app.

3. Tap `Join our open workspace`. Notice the placeholder skeleton appearing for three seconds before server avatar is loaded.

4. Log in to the open workspace and join public channel `test_issue_GH_4810`.

5. Send a message with an image attached. Notice the placeholder skeleton appearing for three seconds before the image is loaded.

6. Send another message in Markdown, for instance:

```markdown
![My image](https://i.imgur.com/KOdB2ed.png)
```

Notice the placeholder skeleton appearing for three seconds before the image is loaded.

NB: if you want to test the behavior on image load failure, comment/uncomment as indicated [here](https://github.com/floels/Rocket.Chat.ReactNative/blob/606ffbfb45a3f7119733b32c75c2f06707273616/app/containers/ImageWithSkeleton/index.tsx#L31-L32) and repeat the actions above. You should see the loading error message:

Capture d’écran 2023-03-16 à 08 48 38

## Screenshots (iOS)

_NB: in the `After` GIFs, the image loading time was artificially set to three seconds, otherwise the effect is not visible._

### Server avatar

| Before | After |
| ------ | ----- |
| Simulator Screen Recording - iPhone 13 - 2023-03-11 at 20 50 39 | Simulator Screen Recording - iPhone 13 - 2023-03-11 at 20 41 47 |

### Image in Markdown message

| Before | After |
| ------ | ----- |
| Simulator Screen Recording - iPhone 13 - 2023-03-11 at 20 47 23 | Simulator Screen Recording - iPhone 13 - 2023-03-11 at 20 39 53 |

### Image in regular message

| Before | After |
| ------ | ----- |
| Simulator Screen Recording - iPhone 13 - 2023-03-11 at 20 46 05 | Simulator Screen Recording - iPhone 13 - 2023-03-11 at 20 37 13 |

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
[NA] I have added necessary documentation (if applicable)
[NA] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by floels** [comment]
@diegolmello ready for your review, let me know what you think.

It is my first PR in this repo and I'm unsure of the QA procedure. I QA-ed only the screens impacted in an iPhone 13 simulator.

**Review by diegolmello — APPROVED** [review]

File: app/containers/ImageSkeletonPlaceholder/index.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import React, { useState } from 'react';
+import { StyleSheet, View } from 'react-native';
+import FastImage, { FastImageProps } from 'react-native-fast-image';
+import SkeletonPlaceholder from 'react-native-skeleton-placeholder';
+
+const ImageSkeletonPlaceholder = ({ style, ...otherProps }: FastImageProps): React.ReactElement => {
```
Comment (diegolmello) [inline app/containers/ImageSkeletonPlaceholder/index.tsx:6]: This naming implies it's just a placeholder, when it's really both image and placeholder/skeleton.
Can you rename it to `ImageWithSkeleton`?
This way is easier to read the code.

File: app/containers/ImageSkeletonPlaceholder/index.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+				/>
+			</SkeletonPlaceholder>
+		</View>
+	);
+
+	return (
+		<View testID='image-skeleton-placeholder'>
+			{isLoading && skeletonPlaceholder}
+			{/* Hide <FastImage /> while it is loading: */}
+			<FastImage style={[{ display: isLoading ? 'none' : 'flex' }, style]} onLoadEnd={handleLoadEnd} {...otherProps} />
```
Comment (diegolmello) [inline app/containers/ImageSkeletonPlaceholder/index.tsx:48]: More of a comment than a requested change, but we need to check `onLoadError` as well to stop showing the placeholder forever.
We can do it on a different PR though, since we don't actually have anything in mind right now.

**Review by diegolmello — CHANGES_REQUESTED** [review]
Almost there.
Thanks for contributing!

File: app/containers/ImageSkeletonPlaceholder/index.tsx (hunk 1 above)
Comment (floels) [reply app/containers/ImageSkeletonPlaceholder/index.tsx]: Good point, fixed ✅

File: app/containers/ImageSkeletonPlaceholder/index.tsx (hunk 2 above)
Comment (floels) [reply app/containers/ImageSkeletonPlaceholder/index.tsx]: Good point. I can add a very basic error message display in this PR. What do you think of the design attached?

Capture d’écran 2023-03-14 à 09 28 13

Comment (diegolmello) [reply app/containers/ImageSkeletonPlaceholder/index.tsx]: We can't do it that way, because depending on the workspace configuration we add userId and token to the url.
Check ShareView on iOS.
Sometimes we need to display a `warning` icon with a message.
Maybe we could do that icon with a simple `Can't display image`.
Just remember to style this text, so it uses font and colors like the rest of the app and we're fine.
I can talk to design team to polish it in the future, if needed.
Thanks!

Comment (diegolmello) [reply app/containers/ImageSkeletonPlaceholder/index.tsx]: @floels

Comment (floels) [reply app/containers/ImageSkeletonPlaceholder/index.tsx]: @diegolmello I added this behavior to the code and tests and I included a screenshot of the resulting design in the PR description.

Let me know what you think!

File: app/containers/ImageWithSkeleton/index.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+			style={{
+				...styles.loadErrorView,
+				backgroundColor: themes[theme].auxiliaryBackground,
+				height: placeholderItemHeight,
+				minHeight: placeholderItemMinHeight
+			}}
+		>
+			<CustomIcon name='warning' color={themes[theme].dangerColor} size={16} />
+			<Text style={{ ...styles.loadErrorText, color: themes[theme].infoText }}>
+				{I18n.isTranslated('Failed_Load_Image') ? I18n.t('Failed_Load_Image') : defaultErrorMessage}
```
Comment (floels) [inline app/containers/ImageWithSkeleton/index.tsx:86]: @diegolmello this requires a new translation key. Could you take care of it, since I don't have access to your Lingohub project?

Comment (diegolmello) [reply app/containers/ImageWithSkeleton/index.tsx:86]: @floels Just add the key on `en` and LingoHub is going to take it.

File: app/containers/ImageSkeletonPlaceholder/index.tsx (hunk 2 above)
Comment (diegolmello) [reply app/containers/ImageSkeletonPlaceholder/index.tsx]: lgtm!

File: app/containers/ImageWithSkeleton/index.tsx (hunk 3 above)
Comment (floels) [reply app/containers/ImageWithSkeleton/index.tsx]: @diegolmello done ✅

**Review by diegolmello — DISMISSED** [review]

**Review by diegolmello — CHANGES_REQUESTED** [review]
@floels this is failing on Detox on Android.
It gets stuck while animating.

https://user-images.githubusercontent.com/804994/226018169-5b21da4d-e0c4-44eb-a67c-17291026c4e1.mp4

**Comment by floels** [comment]
> @floels this is failing on Detox on Android. It gets stuck while animating.

@diegolmello thanks for the feedback.

I am trying to reproduce the issue locally but I don't manage to build the Android app locally. I [posted on the public #react-native channel](https://open.rocket.chat/channel/react-native?msg=KR52CTGjKB2HhQgfW) to ask for help.

**Comment by floels** [comment]
> > @floels this is failing on Detox on Android. It gets stuck while animating.
> 
> @diegolmello thanks for the feedback.
> 
> I am trying to reproduce the issue locally but I don't manage to build the Android app locally. I [posted on the public #react-native channel](https://open.rocket.chat/channel/react-native?msg=KR52CTGjKB2HhQgfW) to ask for help.

@diegolmello I still can't solve this issue when trying to run the Android app locally. I submitted issue #4999 in case other developers have an idea how to solve it. I'll keep you posted.

**Comment by diegolmello** [comment]
> > > @floels this is failing on Detox on Android. It gets stuck while animating.
> > 
> > 
> > @diegolmello thanks for the feedback.
> > I am trying to reproduce the issue locally but I don't manage to build the Android app locally. I [posted on the public #react-native channel](https://open.rocket.chat/channel/react-native?msg=KR52CTGjKB2HhQgfW) to ask for help.
> 
> @diegolmello I still can't solve this issue when trying to run the Android app locally. I submitted issue #4999 in case other developers have an idea how to solve it. I'll keep you posted.

Ok. Hard to deal with it, because we can't reproduce it on our environments or on CI.
We can finish this PR when we have the time.
Let us know if you can fix your Android env.

Don't let this Android issue stop you from contributing though.
If you still want to keep helping, do it for iOS only and we can check Android later 🤞🏼

**Comment by floels** [comment]
> > > > @floels this is failing on Detox on Android. It gets stuck while animating.
> > > 
> > > 
> > > @diegolmello thanks for the feedback.
> > > I am trying to reproduce the issue locally but I don't manage to build the Android app locally. I [posted on the public #react-native channel](https://open.rocket.chat/channel/react-native?msg=KR52CTGjKB2HhQgfW) to ask for help.
> > 
> > 
> > @diegolmello I still can't solve this issue when trying to run the Android app locally. I submitted issue #4999 in case other developers have an idea how to solve it. I'll keep you posted.
> 
> Ok. Hard to deal with it, because we can't reproduce it on our environments or on CI. We can finish this PR when we have the time. Let us know if you can fix your Android env.
> 
> Don't let this Android issue stop you from contributing though. If you still want to keep helping, do it for iOS only and we can check Android later 🤞🏼

Thank you @diegolmello! Indeed if you can take over from here it would be great 😊 

I'll retry to build on Android regularly, hopefully I'll manage to make it work at some point.

Thanks for your help anyhow,

Florian
